node/core/provisioner/src/lib.rs

```diff
@@ -466,18 +469,42 @@ async fn select_candidates(
 	}
 
 	// keep only one candidate with validation code.
-	let mut with_validation_code = false;
-	candidates.retain(|c| {
-		if c.candidate.commitments.new_validation_code.is_some() {
-			if with_validation_code {
-				return false
-			}
+	let mut with_validation_code = None;
+	let mut cumulative_size = 0usize;
+	let mut out = Vec::new();
+
+	// Drain the candidates list to extract a suitable list to propose
+	while let Some(c) = candidates.pop() {
```

Comment:
In theory, we could achieve this via `.into_iter().map(...)`, but I feel like the while let loop is easier to read.